### PR TITLE
libsystem: Extend stream api

### PR DESCRIPTION
**Pull request by feliwir**
- [x] Use streaming api in ZipArchive
- [x] Add Deflate Reader & Writer
- [x] Use Reader & Writer interface for existing io classes

### CODE REVIEW

File: libraries/libfile/ZipArchive.cpp [hunk 1]
```diff
@@ -253,7 +245,12 @@ void ZipArchive::write_entry(const Entry &entry, BinaryWriter &writer, const Vec
 
     writer.put(header);
     writer.put_fixed_len_string(entry.name);
-    writer.put_data(compressed_data.raw_storage(), compressed_data.count());
+
+    uint8_t *data_buffer = nullptr;
+    size_t data_size = 0;
+    compressed.read_all((void **)&data_buffer, &data_size);
+    writer.put_data(data_buffer, data_size);
+    delete[] data_buffer;
```
Comment (sleepy-monax) [inline libraries/libfile/ZipArchive.cpp:253]: Isn't this a use case for copyto or copyfrom ?

**Review by sleepy-monax — CHANGES_REQUESTED** [review]
Again great PR, just one comment :)

File: libraries/libfile/ZipArchive.cpp (hunk 1 above)
Comment (feliwir) [reply libraries/libfile/ZipArchive.cpp:253]: Yeah it should be, but **BinaryWriter** is no real writer yet. I'll fix it, just leave it open

Comment (sleepy-monax) [reply libraries/libfile/ZipArchive.cpp:253]: ok :+1:

**Review by sleepy-monax — APPROVED** [review]

**Pull request merged**